### PR TITLE
introduce dirty list to liveness, eliminate `ins` vector

### PR DESCRIPTION
**Pull request by nikomatsakis**
At least in my measurements, this seems to knock much of the liveness computation off the profile.

r? @Zoxc 
cc @nnethercote

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
Perf is queued.

**Comment by nikomatsakis** [comment]
(I'm updating the broken tests, those are nothing serious, just waiting for the build on my local machine)

**Comment by nikomatsakis** [comment]
Perf URLs will be:

- [comparison](http://perf.rust-lang.org/compare.html?start=2eb6969c6a2efba9e59544bc18b4d96a9f2b2504&end=c27e6f7936761b7b1ff3b3db58ceb9d7fec0cf34&stat=instructions%3Au)

**Comment by nikomatsakis** [comment]
cc https://github.com/rust-lang/rust/issues/51819

**Comment by nikomatsakis** [comment]
cc @rust-lang/compiler — this would be nice to land for NLL (it's a modest gain, but a useful one, and there are several other PRs building on this code). If anybody is able to review, would be appreciated.

File: src/librustc_mir/util/liveness.rs [hunk 1]
```diff
@@ -124,37 +122,35 @@ pub fn liveness_of_locals<'tcx>(mir: &Mir<'tcx>, mode: LivenessMode) -> Liveness
...
+    while let Some(bb) = dirty_queue.pop() {
+        // bits = use ∪ (bits - def)
+        bits.overwrite(&outs[bb]);
+        def_use[bb].apply(&mut bits);
+
+        // add `bits` to the out set for each predecessor; if those
+        // bits were not already present, then enqueue the predecessor
+        // as dirty.
+        for &pred_bb in &predecessors[bb] {
+            if outs[pred_bb].union(&bits) {
```
Comment (Zoxc) [inline src/librustc_mir/util/liveness.rs]: I think there should be a comment indicating that `union` returns `true` if it changes anything. I don't normally associate that with the union operator.

File: src/librustc_mir/util/liveness.rs [hunk 2]
```diff
@@ -124,37 +122,35 @@ pub fn liveness_of_locals<'tcx>(mir: &Mir<'tcx>, mode: LivenessMode) -> Liveness
...
-            // update bits on entry and flag if they have changed
-            if ins[b] != bits {
-                ins[b].overwrite(&bits);
-                changed = true;
+    while let Some(bb) = dirty_queue.pop() {
+        // bits = use ∪ (bits - def)
+        bits.overwrite(&outs[bb]);
+        def_use[bb].apply(&mut bits);
+
+        // add `bits` to the out set for each predecessor; if those
```
Comment (Zoxc) [inline src/librustc_mir/util/liveness.rs]: Maybe add a remark that `bits` now contains the state at the entry point the block (and has the value the previous `ins` variable helt).

File: src/librustc_mir/util/liveness.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc_mir/util/liveness.rs]: Yeah, those methods are undercommented. I'll add some comments. =)

File: src/librustc_mir/util/liveness.rs [hunk 3]
```diff
@@ -438,7 +432,6 @@ pub fn write_mir_fn<'a, 'tcx>(
                 .collect();
             writeln!(w, "{} {{{}}}", prefix, live.join(", "))
         };
-        print(w, "   ", &result.ins)?;
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/librustc_mir/util/liveness.rs]: We already have the code to recompute it -- and in fact in the MIR dump, we include those results. I could update this code to match I suppose.

[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/librustc_mir/util/liveness.rs]: (That is, in the NLL MIR dump)

**Comment by nikomatsakis** [comment]
@bors r=Zoxc

I didn't modify the dumping code because I am running short of time. Easy enough to do in the future.

**Comment by nikomatsakis** [comment]
@bors p=1

Bumping priority as a number of NLL PRs are blocked on this, and NLL perf itself is pretty high priority and time sensitive.

**Pull request merged**